ios/RNGoogleSignin/RNGoogleSignin.m

```diff
@@ -67,14 +103,42 @@ - (dispatch_queue_t)methodQueue
     return dispatch_get_main_queue();
 }
 
+static void RNGoogleSigninRejectWithError(NSString *message, NSError *error, RCTPromiseRejectBlock reject)
+{
+  NSString* errorCode = [NSString stringWithFormat:@"%ld", error.code];
+  NSString* errorMessage = [NSString stringWithFormat:@"RNGoogleSignInError: %@, %@", message, error.description];
+
+  reject(errorCode, errorMessage, error);
+}
+
 - (void)signIn:(GIDSignIn *)signIn didSignInForUser:(GIDGoogleUser *)user withError:(NSError *)error {
+    RCTPromiseResolveBlock resolve = self.isSigningInSilently == YES ? self.currentUserAsyncResolve : self.signInResolve;
+    RCTPromiseRejectBlock reject = self.isSigningInSilently == YES ? self.currentUserAsyncReject : self.signInReject;
 
     if (error != Nil) {
-        return [self.bridge.eventDispatcher sendAppEventWithName:@"RNGoogleSignInError"
-                                                            body:@{
-                                                                   @"message": error.description,
-                                                                   @"code": [NSNumber numberWithInteger: error.code]
-                                                                  }];
+      if (error.code == -1) {
+        RNGoogleSigninRejectWithError(@"Unknown error when signing in.", error, reject);
+      } else if (error.code == -2) {
+        RNGoogleSigninRejectWithError(@"A problem reading or writing to the application keychain.", error, reject);
+      } else if (error.code == -3) {
+        RNGoogleSigninRejectWithError(@"No appropriate applications are installed on the device which can handle sign-in. Both webview and switching to browser have both been disabled.", error, reject);
+      } else if (error.code == -4) {
+        RNGoogleSigninRejectWithError(@"The user has never signed in before with the given scopes, or they have since signed out.", error, reject);
+      } else if (error.code == -5) {
+        RNGoogleSigninRejectWithError(@"The user has canceled the sign in request.", error, reject);
+      } else {
+        // RCTLogError(@"%s: %@ (%d)", __func__, error, error.code);
+        RNGoogleSigninRejectWithError(@"Unknown error and error code when signing in.", error, reject);
+      }
+
+      self.currentUserAsyncResolve = nil;
```

Comment:
the code block starting on line 166 seems to be a copy-paste of this. It should be extracted to avoid duplicity. Going even further, the whole body of this `if` should be in a separate function, along with `RCTPromiseRejectBlock reject = ...`